### PR TITLE
Feature/4769 ting ddb cover client

### PR DESCRIPTION
**Pull request by cableman**
#### Link to issue

https://platform.dandigbib.org/issues/4769

#### Description

Add support for DDB CoverService V2 reset API.

#### Screenshot of the result

Screenshot 2020-06-02 at 10 40 16

#### Checklist

- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

#### Additional comments or questions

Note: This is dependent on https://github.com/ding2/ding2/pull/1617

### CODE REVIEW

File: modules/ting_covers_ddb/includes/ting_covers_ddb.admin.inc [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  $form['ddb']['ting_covers_ddb_enable_logging'] = array(
+    '#type' => 'checkbox',
+    '#title' => t('Enable logging'),
+    '#default_value' => variable_get('ting_covers_ddb_enable_logging', FALSE),
+  );
+
+  $form['ddb']['development'] = array(
+    '#type' => 'fieldset',
+    '#title' => t('Development settings'),
+    '#description' => t('The module request that you have a valid adgangsplatform setup to get an agency token. For testing you can here set an token (which will expire) if you are on an system without Adgangsplatformen enabled. Should <strong>never</strong> be used in production.'),
```
Comment (turegjorup) [inline modules/ting_covers_ddb/includes/ting_covers_ddb.admin.inc:40]: `t('The module request that you have a valid adgangsplatform setup to get an agency token. For testing you can here set an token (which will expire) if you are on an system without Adgangsplatformen enabled. Should <strong>never</strong> be used in production.'),`
->
`t('The module requires that you have a valid "Adgangsplatform" setup to get an agency token. For testing you can manually set an token here if you are on an system without "Adgangsplatformen" enabled. This option should <strong>never</strong> be used in production (Token will not renew once expired). '),`

File: modules/ting_covers_ddb/includes/ting_covers_ddb.admin.inc [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    '#collapsible' => TRUE,
+    '#collapsed' => TRUE,
+  );
+
+  $form['ddb']['development']['token'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Agency token'),
+    '#description' => t('This should be an agency token generated by using the agency af user and password with @ prefixed which you can generate at <a href="https://openplatform.dbc.dk/v3/" target="_blank">https://openplatform.dbc.dk/v3/</a>'),
```
Comment (turegjorup) [inline modules/ting_covers_ddb/includes/ting_covers_ddb.admin.inc:50]: `t('This should be an agency token generated by using the agency af user and password with @ prefixed which you can generate at <a href="https://openplatform.dbc.dk/v3/" target="_blank">https://openplatform.dbc.dk/v3/</a>'),`
->
`t('This should be an agency token generated by using your agency as both user and password with @ prefixed. Token can generated at <a href="https://openplatform.dbc.dk/v3/" target="_blank">https://openplatform.dbc.dk/v3/</a>'),`

File: modules/ting_covers_ddb/ting_covers_ddb.info [hunk 3]
```diff
@@ -0,0 +1,7 @@
+name = Ting covers (DDB CoverService)
+description = Implementing getting covers from DDB CoverService
```
Comment (turegjorup) [inline modules/ting_covers_ddb/ting_covers_ddb.info:2]: Implementing -> Implements

File: modules/ting_covers_ddb/ting_covers_ddb.module [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  $config->setHost(_ting_covers_ddb_get_url());
+
+  return $config;
+}
+
+/**
+ * Get access token.
+ *
+ * If development agency token have been set under the administration for this
+ * module. Use it else use the library's agency token.
```
Comment (turegjorup) [inline modules/ting_covers_ddb/ting_covers_ddb.module:115]: `If development agency token have been set under the administration for this module. Use it else use the library's agency token.`
->
`Return the development agency token if set for this module. Else return the library's agency token.`

File: modules/ting_covers_ddb/includes/ting_covers_ddb.admin.inc (hunk 1 above)
Comment (cableman) [reply modules/ting_covers_ddb/includes/ting_covers_ddb.admin.inc:40]: Updated the comment.

File: modules/ting_covers_ddb/includes/ting_covers_ddb.admin.inc (hunk 2 above)
Comment (cableman) [reply modules/ting_covers_ddb/includes/ting_covers_ddb.admin.inc:50]: Updated the comment.

File: modules/ting_covers_ddb/ting_covers_ddb.info (hunk 3 above)
Comment (cableman) [reply modules/ting_covers_ddb/ting_covers_ddb.info:2]: Fixed

File: modules/ting_covers_ddb/ting_covers_ddb.module (hunk 4 above)
Comment (cableman) [reply modules/ting_covers_ddb/ting_covers_ddb.module:118]: Updated the comment.

File: modules/ting_covers_ddb/includes/ting_covers_ddb.admin.inc [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    '#collapsed' => TRUE,
+  );
+
+  $default = variable_get('ting_covers_ddb_development_token', FALSE);
+  $form['ddb']['development']['ting_covers_ddb_development_token'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Agency token'),
+    '#default_value' => $default ? $default : '',
+    '#description' => t('This should be an agency token generated by using your agency as both user and password with @ prefixed. Token can generated at <a href="https://openplatform.dbc.dk/v3/" target="_blank">https://openplatform.dbc.dk/v3/</a>'),
+    '#collapsible' => TRUE,
```
[the remark this answers is not in the document]
Comment (cableman) [reply modules/ting_covers_ddb/includes/ting_covers_ddb.admin.inc:51]: Fixed :-)

File: modules/ting_covers_ddb/ting_covers_ddb.module [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      // image that is large enough to scale down to that size. So we fallback
+      // to original image from the service.
+      $source_url = $source_fallback_url;
+    }
+
+    // Return the path to the cover.
+    $covers[$cover->getId()] = $source_url;
+  }
+
+  _ting_covers_ddb_debug_log('Found covers for pids: ' . implode(', ', array_keys($covers)));
```
[the remark this answers is not in the document]
Comment (cableman) [reply modules/ting_covers_ddb/ting_covers_ddb.module:89]: Added more logging.

**Comment by JrgenGNielsen** [comment]
reviewed and approved. Behat test error don't seem to be related to this PR

File: modules/ting_covers_ddb/ting_covers_ddb.module [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+/**
+ * Helper to get configuration for the service.
+ *
+ * @return \CoverService\Configuration
+ *   API client configuration.
+ *
+ * @throws \DingProviderDoesntImplement
+ * @throws \DingProviderNoProvider
+ */
+function _ting_cover_ddb_get_clint_config() {
```
Comment (holt83) [inline modules/ting_covers_ddb/ting_covers_ddb.module:105]: Typo?

**Pull request closed without merging**